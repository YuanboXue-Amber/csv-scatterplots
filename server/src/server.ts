import express from 'express';
import multer from 'multer';
import cors from 'cors';

const server = express();
server.use(cors());

// store file in public folder with name date+originalName
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, 'public');
  },
  filename: (req, file, cb) => {
    cb(null, Date.now() + '-' + file.originalname);
  },
});
const upload = multer({ storage }).single('file');

server.get('/', (req: any, res: any) => {
  return res.send('Hello Server');
});

server.post('/upload', (req: any, res: any) => {
  upload(req, res, (err) => {
    if (err) {
      console.error(`Error uploading file ${err}`);
      return res.status(500).json(err);
    }

    if (req.file) {
      console.log(`File uploaded to ${req.file.filename}`);
      return res.status(200).send(req.file);
    }

    return res.status(500);
  });
});

server.listen(8000, () => {
  console.log('Backend server running on port 8000');
});
