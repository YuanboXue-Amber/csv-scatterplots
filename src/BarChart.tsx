import React, { Component } from 'react';
import * as d3 from 'd3';
// tslint:disable-next-line: no-var-requires
const dsv = require('d3-dsv');

export class BarChart extends Component<{}, {}> {
  componentDidMount() {
    this.draw();
  }

  async fetchFile() {
    const response = await fetch(
      'https://raw.githubusercontent.com/thegazettedata/d3charts/master/projects/cr-bike-trails/data/letter-frequency.csv',

    );
    // Check if the request is 200
    if (response.ok) {
      return response.text();
    }
    return Promise.reject(response);
  }

  // Draw BarChart on div with id "BarChart", replace if it exists
  async draw() {
    // const data = [4, 8, 15, 16, 23, 42];
    // const csvString =
    // `name,value\nLocke,4\nReyes,8\nFord,15\nJarrah,16\nShephard,23\nKwon,42`;

    const csvString = await this.fetchFile();

    const rawData = d3.csvParse(csvString, dsv.autoType);
    const data = rawData.map<{ name: string; value: number }>(a => {
      return { name: a[rawData.columns[0]], value: a[rawData.columns[1]] };
    });

    const width = 800;
    const height = 600;
    const svg = d3
      .select('#BarChart')
      .selectAll('svg')
      .data([1])
      .join('svg');
    svg
      .attr('height', height)
      .attr('width', width);

    const margin = {top: 20, left: 30, bottom: 20, right: 20};

    const xscale = d3
      .scaleBand()
      .domain(data.map(a => a.name))
      .range([margin.left, width - margin.right])
      .padding(0.1);

    const yscale = d3
      .scaleLinear()
      .domain([0, (d3.max(data.map(a => a.value)) ?? 0 ) * 1.2])
      .range([height - margin.bottom, margin.top]);

    const xAxis = (g: any) => { g
        .append('g')
      .attr('transform', `translate(0, ${height - margin.bottom})`)
      .call(d3.axisBottom(xscale));
    };

    const yAxis = (g: any) => { g
        .append('g')
      .attr('transform', `translate(${margin.left}, 0)`)
      .call(d3.axisLeft(yscale).tickFormat(d => `${(d.valueOf() * 100).toFixed(0)}%`));
    };

    const ytitle = (g: any) => { g
        .append('text')
      .text('Frequency ^')
      .attr('font-size', xscale.bandwidth() / 2)
      .attr('transform', `translate(0, ${xscale.bandwidth() / 2})`);
    };

    svg.call(xAxis);
    svg.call(yAxis);
    svg.call(ytitle);

    const gbars = svg
      .append('g')
    .selectAll('g')
    .data(data)
    .join('g');

    gbars
        .append('rect')
      .style('fill', 'steelblue')
      .attr('height', d => yscale(0) - yscale(d.value))
      .attr('width', xscale.bandwidth())
      .attr('x', d => xscale(d.name) ?? 0)
      .attr('y', d => yscale(d.value));

    gbars
      .append('text')
      .attr('font-size', xscale.bandwidth() / 2)
      .text(d => `${(d.value * 100).toFixed(2)}%`)
      .attr('x', d => xscale(d.name) ?? 0)
      .attr('y', d => yscale(d.value) - 2);
  }

  render() {
    return <div id="BarChart"/>;
  }
}
