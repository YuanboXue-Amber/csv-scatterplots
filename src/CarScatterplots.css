.dots circle {
  opacity: 0.5;
  fill: #f2669e;
}

.tick line {
  stroke: #C0C0BB;
}

text {
  font-family: sans-serif;
  fill: #635F5D;
}

.tick text {
  font-size: 1.5em;
}

.axisTitle text {
  font-size: 2em;
}

.chart-title text {
  font-size: 3em;
  fill: black;
}