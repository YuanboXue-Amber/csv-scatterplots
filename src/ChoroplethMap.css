#ChoroplethMap {
  padding-top: 3em;
}

.sphere {
  fill: #3bb9b9bd;
}

.country {
  /* fill: #b5f16b; */
  stroke: black;
  stroke-width: 0.15px;
}

.country:hover {
  fill: #e4747e;
}

.hilighted {
  stroke-width: 0.5px;
}