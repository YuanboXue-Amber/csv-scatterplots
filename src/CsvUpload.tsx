import React, { Component } from 'react';
import * as d3 from 'd3';
import { isNullOrUndefined } from 'util';
import { ScatterPlots } from './ScatterPlots';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.min.css';

interface CsvUploadState { selectedFile: any; csvData: number[][]; }

export class CsvUpload extends Component<{}, CsvUploadState> {

  // Check for the various File API support.
  componentDidMount() {
    if (window.File && window.FileReader && window.Blob) {
      // all the File APIs are supported.
    } else {
      alert('The File APIs are not fully supported in this browser.');
    }
  }

  constructor(props: Readonly<{}>) {
    super(props);
    this.state = {
      selectedFile: null,
      csvData: [],
    };
    this.readCsv = this.readCsv.bind(this);
    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
  }

  isNumber(value: string | undefined): boolean {
     return ((value != null) &&
             !isNaN(Number(value.toString())));
  }

  // Read uploaded file, parse with d3.csv, and set this.state.csvData
  readCsv(file: Blob) {
    const reader = new FileReader();
    reader.onloadend = (event) => {
      if (!isNullOrUndefined(event.target)) {
        const csvUrl = event.target.result as string; // get content of the uploaded file
        d3.csv(csvUrl).then( // read file as csv
          (data) => {
            if (data.length === 0) {
              toast.error(`Uploaded file is empty`);
            } else {
              const cols = data.columns;
              if (cols.length < 2) {
                toast.error(`CSV file needs to have at least 2 columns`);
              } else {
                // read csv file and set this.state.csvData
                const col1 = Object.keys(data[0])[0];
                const col2 = Object.keys(data[0])[1];
                const csvData: number[][] = [];
                let invalidValueFound = false;
                data.forEach((d) => {
                  const numx = d[col1];
                  const numy = d[col2];
                  if (this.isNumber(numx) && this.isNumber(numy)) {
                    csvData.push([Number(numx), Number(numy)]);
                  } else {
                    invalidValueFound = true;
                  }
                });
                if (invalidValueFound) {
                  toast.warn(`CSV fields should contain only numbers. Invalid values are ignored.`);
                }
                this.setState({
                  selectedFile: this.state.selectedFile,
                  csvData,
                });
              }
            }
          },
          (err) => {
            toast.error(`Error reading uploaded file: ${err}`);
          });
      }
    };
    reader.readAsDataURL(file);
  }

  // Validate that the uploaded file has correct type
  onChangeHandler(event: any) {
    const file = event.target.files[0];
    if (file.type === 'text/plain' || file.type === 'application/vnd.ms-excel' || file.type === 'text/csv') {
      this.setState({
        selectedFile: file,
      });
    } else {
      toast.error(`${file.type} is not a supported format. Only text/plain is supported`);
    }
  }

  onClickHandler() {
    const file = this.state.selectedFile;
    if (isNullOrUndefined(file)) { return; }
    this.readCsv(file);
  }

  render() {
    return (
      <div className="container">
        <div className="row">
          <div className="offset-md-3 col-md-6">
            <div className="form-group files">
              <label>Upload Your CSV File </label>
              <a href="sample.csv" download="sample.csv">
                <button type="button" className="btn btn-info btn-sm float-right">
                  Download a sample CSV file
                </button>
              </a>
              <input type="file" className="form-control" onChange={this.onChangeHandler} />
            </div>
            <div className="form-group">
              <ToastContainer className="toast-top-right" />
            </div>

            <button type="button" className="btn btn-success btn-block" onClick={this.onClickHandler}>
              Upload
            </button>
            <ScatterPlots data={this.state.csvData} />
          </div>
        </div>
      </div>
    );
  }
}
