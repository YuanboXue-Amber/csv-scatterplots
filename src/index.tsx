import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
// import { BarChart } from './BarChart';
// import { RandomLetters } from './RandomLetters';
// import { CarScatterplots } from './CarScatterplots';
// import { TemperatureLineArea } from './TemperatureLineArea';
// import { FruitBowl } from './FruitBowl';
import { WorldMap } from './WorldMap';

ReactDOM.render(
  <div className="container">
    {/* <div className="container">
      <BarChart />
    </ div>
    <div className="container">
      <RandomLetters />
    </ div> */}
    {/* <div className="container">
      <CarScatterplots />
    </ div> */}
    {/* <div className="container">
      <TemperatureLineArea />
    </ div> */}
    {/* <div className="container">
      <FruitBowl />
    </ div> */}
    <div className="container">
      <WorldMap />
    </ div>
  </ div>,
  document.getElementById('root'),
);
