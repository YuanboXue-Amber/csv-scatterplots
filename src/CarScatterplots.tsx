import React, { Component } from 'react';
import * as d3 from 'd3';
import './CarScatterplots.css';
// tslint:disable-next-line: no-var-requires
const dsv = require('d3-dsv');

interface ICar {
  mpg: number;
  cylinders: number;
  displacement: number;
  horsepower: number;
  weight: number;
  acceleration: number;
  year: number;
  origin: string;
  name: string;
}

export class CarScatterplots extends Component<{}, {}> {
  componentDidMount() {
    this.draw();
  }

  async fetchFile() {
    const response = await fetch(
      'https://vizhub.com/curran/datasets/auto-mpg.csv',
    );
    // Check if the request is 200
    if (response.ok) {
      return response.text();
    }
    return Promise.reject(response);
  }

  async draw() {

    const csvString = await this.fetchFile();
    const data: d3.DSVParsedArray<ICar> = d3.csvParse(csvString, dsv.autoType);

    const title = 'Cars';
    const xvar = (d: ICar) => d.weight;
    const yvar = (d: ICar) => d.horsepower;
    const xinfo = 'weight';
    const yinfo = 'horsepower';
    const width = 960;
    const height = 500;
    const margin = {top: 80, left: 80, bottom: 80, right: 60};
    const innerwidth = width - margin.left - margin.right;
    const innerheight = height - margin.top - margin.bottom;
    const radius = 5;

    const svg = d3
      .select('#CarScatterplots')
      .selectAll('svg')
      .data([1])
      .join('svg')
        .attr('height', height)
        .attr('width', width);

    const xscale = d3
      .scaleLinear()
      .domain([d3.min(data, xvar) ?? 0, d3.max(data, xvar) ?? 0])
      .range([0, innerwidth])
      .nice();
    const yscale = d3
      .scaleLinear()
      .domain([d3.min(data, yvar) ?? 0, d3.max(data, yvar) ?? 0])
      .range([innerheight, 0])
      .nice();

    const xAxis = (g: any) => {
      const xAxisG = g.append('g');
      xAxisG
        .attr('transform', `translate(${margin.left}, ${margin.top + innerheight})`)
        .call(d3.axisBottom(xscale)
                .tickSizeInner(0 - innerheight)
                .tickPadding(10));
      xAxisG
        .select('.domain')
        .remove();
    };
    const yAxis = (g: any) => {
      const yAxisG = g.append('g');
      yAxisG
        .attr('transform', `translate(${margin.left}, ${margin.top})`)
        .call(d3.axisLeft(yscale)
                .tickSizeInner(0 - innerwidth)
                .tickPadding(5));
      yAxisG
        .select('.domain')
        .remove();
    };

    const xTitle = (g: any) => {
      const xTitleG = g.append('g');
      xTitleG
        .attr('class', 'axisTitle')
        .attr('transform', `translate(${margin.left + innerwidth / 2}, ${margin.top + innerheight + 50})`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(xinfo);
    };
    const yTitle = (g: any) => {
      const yTitleG = g.append('g');
      yTitleG
        .attr('class', 'axisTitle')
        .attr('transform', `translate(40, ${margin.top + innerheight / 2}) rotate(-90)`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(yinfo);
    };

    const chartTitle = (g: any) => {
      const titleG = g.append('g');
      titleG
        .attr('class', 'chart-title')
        .attr('transform', `translate(${margin.left + innerwidth / 2}, 50)`)
        .attr('text-anchor', 'middle')
        .append('text')
          .text(title);
    };

    svg.call(xAxis);
    svg.call(yAxis);
    svg.call(xTitle);
    svg.call(yTitle);
    svg.call(chartTitle);

    const dots = svg
      .append('g')
        .attr('class', 'dots')
        .selectAll('circle')
        .data(data)
        .join('circle')
        .attr('r', radius)
        .attr('cx', d => xscale(xvar(d)) + margin.left)
        .attr('cy', d => yscale(yvar(d)) + margin.top);
    const tooltip = svg.append('g')
      .attr('class', 'mytooltip')
      .append('text')
      .style('opacity', 0);
    dots
      .on('mouseover', function (d) {
        tooltip
          .transition()
          .style('opacity', 1);
        tooltip
          .attr('x', d3.select(this).attr('cx'))
          .attr('y', d3.select(this).attr('cy'))
          .text(`(${xvar(d)}, ${yvar(d)})`);
      })
      .on('mouseout', (d) => {
        tooltip
          .transition()
          .style('opacity', 0);
      });
  }

  render() {
    return <div id="CarScatterplots"/>;
  }
}
