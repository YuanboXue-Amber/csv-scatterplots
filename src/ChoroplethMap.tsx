import React, { Component } from 'react';
import * as d3 from 'd3';
import './ChoroplethMap.css';
import { colorLegendVertical } from './ColorLegendVertical';
// tslint:disable-next-line: no-var-requires
const topojson = require('topojson-client');

export class ChoroplethMap extends Component<{}, {}> {
  worldData: any[] = [];

  componentDidMount() {
    this.draw();
  }

  async getWorldData() {
    const [worldTopo, worldSpec]  = await Promise.all([
      d3.json('https://unpkg.com/world-atlas@1.1.4/world/50m.json'),
      d3.tsv('https://unpkg.com/world-atlas@1.1.4/world/50m.tsv'),
    ]);

    // worldGeo.id is country. worldGeo.geometry contains type Polygon and coordinates array
    const worldGeo = (topojson.feature(worldTopo, worldTopo.objects.countries)).features;

    // process both data to get worldData as:
    // worldGeo
    // worldGeo.name = worldSpec.name, worldGeo.spec = worldSpec.economy
    const map = new Map();
    worldGeo.forEach((country: any) => {
      map.set(country.id, country);
    });

    worldSpec.forEach((country: any) => {
      const countryGeo = map.get(country.iso_n3);
      countryGeo.name = country.name;
      countryGeo.spec = country.economy; // pick economy as the 'spec' we are interested in
      map.set(country.iso_n3, countryGeo);
    });
    this.worldData = Array.from(map.values());
  }

  // Draw ChoroplethMap on div with id "ChoroplethMap", replace if it exists
  async draw() {

    await this.getWorldData();

    const width = 960;
    const height = 800;
    const svg = d3
      .select('#ChoroplethMap')
      .selectAll('svg')
      .data([null])
      .join('svg')
        .attr('height', height)
        .attr('width', width);

    // map generator
    const projection = d3.geoNaturalEarth1();
    const pathGenerator = d3.geoPath(projection);

    // color scale
    const colorScale = d3.scaleOrdinal()
      .domain(this.worldData.map((d: any) => d.spec).sort());
    colorScale
      .range(d3.schemeSpectral[colorScale.domain().length]);

    // draw stable part of the map
    const mapG = svg
      .selectAll('.map')
      .data([null])
      .join('g')
        .attr('class', 'map');

    mapG
      .selectAll('.sphere')
      .data([null])
      .join('path')
        .attr('class', 'sphere')
        .attr('d', d => pathGenerator({type: 'Sphere'}));

    // zoom
    const zoom = d3.zoom()
      .scaleExtent([1, 8])
      .on('zoom', () => {
        mapG.attr('transform', d3.event.transform);
      });
    svg.call(zoom as any); // somehow, when it is mapG.call, panning by drag became really hard

    // draw color legend
    const colorLegendG = svg
      .selectAll('.colorLegend')
      .data([null])
      .join('g')
        .attr('class', 'colorLegend')
        .attr('transform', 'translate(0, 250)');
    const selector = colorLegendG;
    const colorRadius = 10;
    const colorDistance = 5;

    // add event when click color legend
    let clickedDomain: any;
    const onClick = (clicked: string) => {
      console.log(clicked);
      clickedDomain = clicked;
      renderMap();
    };

    // render dynamic part of map
    const renderMap = () => {
      const countries = mapG
        .selectAll('.country')
        .data(this.worldData)
        .join('path')
          .attr('class', 'country')
          .attr('d', (d: any) => pathGenerator(d))
          .attr('fill', (d: any) => colorScale(d.spec) as string)
          .attr('opacity', (d: any) =>
            (!clickedDomain || d.spec === clickedDomain)
            ? 1
            : 0.2,
          )
          .classed('hilighted', d => d.spec === clickedDomain);

      countries
        .selectAll('title')
        .data(d => [d])
        .join('title')
          .text((d: any) => `${d.name}: ${d.spec}`); // set hover text

      colorLegendVertical({ selector, colorScale, colorRadius, colorDistance, textwidth: 200, onClick, clickedDomain});
    };

    renderMap();
  }

  render() {
    return <div id="ChoroplethMap"/>;
  }
}
