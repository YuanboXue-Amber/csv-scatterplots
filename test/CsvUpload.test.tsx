import React from 'react';
import { render } from '@testing-library/react';
import { CsvUpload } from '../src/CsvUpload';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test('CsvUpload has componentDidMount triggered', () => {
  const spy = jest.spyOn(CsvUpload.prototype, 'componentDidMount');
  shallow(<CsvUpload />);
  expect(spy).toHaveBeenCalled();
});

test('CsvUpload renders correct html', () => {
  const csvUpload = render(<CsvUpload />);
  expect(csvUpload.container).toMatchSnapshot();
});

describe('CsvUpload calls onChangeHandler when file is uploaded.', () => {
  let component: any;
  beforeEach(() => {
      component = shallow(<CsvUpload />);
  });

  test('Image uploaded, state.selectedFile should be undefined', () => {
    const event = {
      preventDefault() {},
      target: {
        files: [
          {type: 'image/jpeg'},
        ],
      },
    };
    component.find('input').simulate('change', event);
    expect(component.state()).toMatchSnapshot();
  });

  test('Text file uploaded, state.selectedFile should be changed', () => {
    const event = {
      preventDefault() {},
      target: {
        files: [
          {type: 'text/plain'},
        ],
      },
    };
    component.find('input').simulate('change', event);
    expect(component.state()).toMatchSnapshot();
  });
});

test('CsvUpload calls onClickHandler when upload buttom is clicked', () => {
  jest.spyOn(CsvUpload.prototype, 'onClickHandler');

  const component = shallow(<CsvUpload />);
  const file = new Blob([''], {type: 'text/plain;charset=ISO-8859-1'});
  component.setState({ selectedFile: file, csvData: [] });
  const event = {
    preventDefault() {},
    target: {
      files: [file],
    },
  };

  const buttons = component.find('button');
  expect(buttons.length).toBe(2);

  const uploadButton = component.find('button').at(1);
  uploadButton.simulate('click', event);
  expect(CsvUpload.prototype.onClickHandler).toHaveBeenCalledWith(event);
});
