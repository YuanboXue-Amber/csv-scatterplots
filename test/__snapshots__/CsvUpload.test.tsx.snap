// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvUpload calls onChangeHandler when file is uploaded. Image uploaded, state.selectedFile should be undefined 1`] = `
Object {
  "csvData": Array [],
  "selectedFile": null,
}
`;

exports[`CsvUpload calls onChangeHandler when file is uploaded. Text file uploaded, state.selectedFile should be changed 1`] = `
Object {
  "csvData": Array [],
  "selectedFile": Object {
    "type": "text/plain",
  },
}
`;

exports[`CsvUpload renders correct html 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="offset-md-3 col-md-6"
      >
        <div
          class="form-group files"
        >
          <label>
            Upload Your CSV File 
          </label>
          <a
            download="sample.csv"
            href="sample.csv"
          >
            <button
              class="btn btn-info btn-sm float-right"
              type="button"
            >
              Download a sample CSV file
            </button>
          </a>
          <input
            class="form-control"
            type="file"
          />
        </div>
        <div
          class="form-group"
        >
          <div
            class="Toastify"
          />
        </div>
        <button
          class="btn btn-success btn-block"
          type="button"
        >
          Upload
        </button>
        <div
          id="scatterPlots"
        >
          <svg
            align="center"
            height="500"
            width="600"
          >
            <g
              class="axis"
              fill="none"
              font-family="sans-serif"
              font-size="10"
              text-anchor="middle"
              transform="translate(100, 400)"
            >
              <path
                class="domain"
                d="M0.5,6V0.5H300.5V6"
                stroke="currentColor"
              />
              <g
                class="tick"
                opacity="1"
                transform="translate(150.5,0)"
              >
                <line
                  stroke="currentColor"
                  y2="6"
                />
                <text
                  dy="0.71em"
                  fill="currentColor"
                  y="9"
                >
                  0
                </text>
              </g>
            </g>
            <g
              class="axis"
              fill="none"
              font-family="sans-serif"
              font-size="10"
              text-anchor="end"
              transform="translate(100, 100)"
            >
              <path
                class="domain"
                d="M-6,300.5H0.5V0.5H-6"
                stroke="currentColor"
              />
              <g
                class="tick"
                opacity="1"
                transform="translate(0,150.5)"
              >
                <line
                  stroke="currentColor"
                  x2="-6"
                />
                <text
                  dy="0.32em"
                  fill="currentColor"
                  x="-9"
                >
                  0
                </text>
              </g>
            </g>
          </svg>
          <div
            class="tooltip"
            style="opacity: 0;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
